core/src/retransmit_stage.rs

```diff
@@ -209,49 +213,76 @@ pub type ShredFilter = LruCache<(Slot, u32, bool), Vec<u64>>;
 
 pub type ShredFilterAndHasher = (ShredFilter, PacketHasher);
 
-// Return true if shred is already received and should skip retransmit
+// Returns None if shred is already received and should skip retransmit.
+// Otherwise returns shred's slot.
 fn check_if_already_received(
     packet: &Packet,
-    shreds_received: &Arc<Mutex<ShredFilterAndHasher>>,
-) -> bool {
-    match get_shred_slot_index_type(packet, &mut ShredFetchStats::default()) {
-        Some(slot_index) => {
-            let mut received = shreds_received.lock().unwrap();
-            let hasher = received.1.clone();
-            if let Some(sent) = received.0.get_mut(&slot_index) {
-                if sent.len() < MAX_DUPLICATE_COUNT {
-                    let hash = hasher.hash_packet(packet);
-                    if sent.contains(&hash) {
-                        return true;
-                    }
-
-                    sent.push(hash);
-                } else {
-                    return true;
-                }
+    shreds_received: &Mutex<ShredFilterAndHasher>,
+) -> Option<Slot> {
+    let shred = get_shred_slot_index_type(packet, &mut ShredFetchStats::default())?;
+    let mut shreds_received = shreds_received.lock().unwrap();
+    let (cache, hasher) = shreds_received.deref_mut();
+    match cache.get_mut(&shred) {
+        Some(sent) if sent.len() >= MAX_DUPLICATE_COUNT => None,
+        Some(sent) => {
+            let hash = hasher.hash_packet(packet);
+            if sent.contains(&hash) {
+                None
             } else {
-                let hash = hasher.hash_packet(&packet);
-                received.0.put(slot_index, vec![hash]);
+                sent.push(hash);
+                Some(shred.0)
             }
+        }
+        None => {
+            let hash = hasher.hash_packet(packet);
+            cache.put(shred, vec![hash]);
+            Some(shred.0)
+        }
+    }
+}
 
-            false
+// Returns true if turbine retransmit peers patch (#14565) is enabled.
+fn enable_turbine_retransmit_peers_patch(
+    shred_slot: Slot,
+    bank_forks: &RwLock<BankForks>,
+    cache: &mut HashMap<Slot, Arc<Bank>>,
+) -> bool {
+    let bank: Arc<Bank> = cache
+        .entry(shred_slot)
+        .or_insert_with(|| {
+            let bank_forks = bank_forks.read().unwrap();
+            match bank_forks.get(shred_slot) {
+                Some(bank) => bank.clone(),
+                None => bank_forks.root_bank(),
+            }
+        })
+        .clone();
+    let feature_slot = bank
```

Comment:
On testnet, this feature is effectively active now right?  If so then this appears to turn the feature off for testnet nodes as they update to this patch, so I think we'd want to either add a check for !Testnet here unfortunately

Reply:
so i was thinking not to backport this to testnet,
or when backporting modify it so that it is effectively a no-op

Reply:
> so i was thinking not to backport this to testnet,

What does that mean, not backport to the v1.5 branch?  That doesn't work since the 1.5 branch has to work with both mainnet-beta and testnet

Reply:
I see.

How about just rolling it out to testnet and taking the risk that it **might** break testnet during the upgrade and we may need a restart?
The `DATA_PLANE_FANOUT` is 200 and so there is a good chance that the cluster keeps working with or without https://github.com/solana-labs/solana/pull/14565 during the course of upgrade (because there will be only 2 layers, and there will be a spanning tree nevertheless).

The advantage is that we can test dynamically enabling the feature on testnet before doing so on mainnet.

Reply:
Sure.  Restarting testnet is always a good exercise :)